lib/types/jtd-schema.ts

```diff
@@ -0,0 +1,103 @@
+/** required keys of an object, not undefined */
+type RequiredKeys<T> = {
+  [K in keyof T]-?: undefined extends T[K] ? never : K
+}[keyof T]
+
+/** optional or undifined-able keys of an object */
+type OptionalKeys<T> = {
+  [K in keyof T]-?: undefined extends T[K] ? K : never
+}[keyof T]
+
+/** type is true if all elements of T extend T */
+type AllExtend_<T, E> = T extends E ? true : false
+type AllExtend<T, E> = false extends AllExtend_<T, E> ? false : true
+
+/** type is true if type is identically string */
+type IsString_<T> = T extends string ? (string extends T ? true : false) : false
```

Comment:
I am wondering why the second line is necessary here? The first line seems to say that for T to be identically string, T should extend string and string should extend T - which I thought would be enough, but it is probably not correct given that the second line is needed... And I think for majority of T the first line is enough, but there are some Ts when it is not - what are they?

As a side comment - Typescript is weird really if you just can have type level equality without these trick :)

AllExtend I understand even less, but it looks similar - what is the example there?

Maybe it is worth defining a utility type `TypeEquality<T1, T2>` that would only evaluate to true only if T1 is identical to T2 - although I am not sure if what happens here with string can be generalised - maybe not...

Reply:
The answer ties to a lot of the choices I've made, and it's a design decision, one I'm open to having. In this case it has to do with how typescript handles `X extends Y ? ... : ...` when X is a union. In that case typescript distributes the conditional across every element of the union (see [distributive conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types)). In this case, the decision that I've made is that that the `JDTSchema<number | string>` shouldn't create a valid schema. However, if you write `T extends string ? string extends T ? MyType<T> : never : never` of `string | number` because of distributive conditional types, that will result in `string | never` which is just `string`. Thus, this solution uses this trick to assert that the type isn't a union with string as a member, but is specifically only the type string.

Yes, it is a little weird, and mostly has to do with the way they wrote the conditional type system. I think the general principle is that typescript is meant to mimic javascript, so you care more about extending (e.g. having the relevant properties) than you care about strict matching, e.g. only having the specific properties.

All extend is similar but instead is saying, every element of the union extends the type. This way `AllExtend<1 | 2 | 3, number>` is still true, and `AllExtend<true, bool>` is still true. This is "necessary" because JTD only defines enums for strings, not for numbers or bools, so we need to extend them. Arguably, `AllExtend` is an interesting design decision. I could see the argument that `JDTSchemaType` should reject that typing because JTD can't actually verify that type.

TypeEquality should work in place of IsString, so I can change that. Depending on the answer / discussion about the above question, they might all become type equality.

Finally, for simple types like this, the typescript playground is a good place to test out how things are resolving.

Reply:
Thanks a lot - got it - need to get into habit of using the playground indeed.

>  I could see the argument that JDTSchemaType should reject that typing because JTD can't actually verify that type.

Interesting. I think the current trade-off is better - to align JTD Schema with the existing type, rather than to restrict type to JTD schema capabilities (there is type code generation approach for the latter). We could add JTDSchemaStrict that would be stricter in that regard